### PR TITLE
Sasha pasha branch

### PR DESCRIPTION
**Pull request by vshcherb**

### CODE REVIEW

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+	private TrkSegment before = new TrkSegment();
+	// cache should be deleted if after changed or snappedToRoadPoints
+	private TrkSegment beforeCacheForSnap;
+
+	private TrkSegment after = new TrkSegment();
+	// cache should be deleted if after changed or snappedToRoadPoints
+	private TrkSegment afterCacheForSnap;
+
+	private boolean inMovePointMode;
+	private boolean inAddPointBeforeMode;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java]: This variables pobably needs to be deleted "inAddPointBeforeMode", "inAddPointAfterMode" - doesn't seem to be usable

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+public class MeasurementEditingContext {
+
+	private MapActivity mapActivity;
+
+	private TrkSegment before = new TrkSegment();
+	// cache should be deleted if after changed or snappedToRoadPoints
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java]: "if after" -> "if before"

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+public class MeasurementEditingContext {
+
+	private MapActivity mapActivity;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java]: It looks a very very wrong place for MapActivity. This class is data not UI

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+	private Queue<Pair<WptPt, WptPt>> snapToRoadPairsToCalculate = new LinkedList<>();
+	private Map<Pair<WptPt, WptPt>, List<WptPt>> snappedToRoadPoints = new HashMap<>();
+
+	private List<WptPt> measurementPoints = new LinkedList<>();
+
+	public void setMapActivity(MapActivity mapActivity) {
+		this.mapActivity = mapActivity;
+	}
+
+	public void setBefore(TrkSegment before) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java]: These methods should not be accessible from outside, except possibly constructure. All modifications should be done in this class.

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+				}
+			}
+		} else {
+			before.points.addAll(measurementPoints);
+		}
+		addBeforeRenders();
+		after = new TrkSegment();
+	}
+
+	public void recreateSegments(int position, boolean addAfter, boolean addBefore) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java]: 2 methods recreateSegments - looks wrong

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java [hunk 6]
```diff
@@ -123,16 +110,15 @@
 	private boolean gpxPointsAdded;
 	private ApplicationMode snapToRoadAppMode;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java]: private boolean gpxPointsAdded;
  private ApplicationMode snapToRoadAppMode;

This should be deleted

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java [hunk 7]
```diff
@@ -123,16 +110,15 @@
...
 
-	private boolean inMovePointMode;
-	private boolean inAddPointAfterMode;
-	private boolean inAddPointBeforeMode;
 	private boolean isInSnapToRoadMode;
 
 	private int selectedPointPos = -1;
 	private WptPt selectedCachedPoint;
 
 	private int positionToAddPoint = -1;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java]: Should be deleted (already in context)
private boolean isInSnapToRoadMode;
private int selectedPointPos = -1;
  	private WptPt selectedCachedPoint;
  
  	private int positionToAddPoint = -1;

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java (hunk 6 above)
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java]: Should be also in context
private NewGpxData newGpxData;

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java (hunk 2 above)
Comment (AlexSyt) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java]: fixed

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java (hunk 3 above)
Comment (AlexSyt) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementEditingContext.java]: fixed

File: OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java (hunk 6 above)
Comment (AlexSyt) [inline OsmAnd/src/net/osmand/plus/measurementtool/MeasurementToolFragment.java]: snapToRoadAppMode deleted

**Pull request closed without merging**